### DISCUSSION TITLE
[Item/Text] 适合艺术文科类文献阅读笔记关联注释

### DISCUSSION THREAD

**Discussion by swimpinkNo1** [windingwind/zotero-better-notes#1165] [Note Templates]
### Description

适合苹果系统的，文科艺术类阅读或文献笔记，主打一个高效和高颜值。
我的 zotero 版本是 7.07 MacBook 的系统是14.6.1 (23G93)
用不同的颜色整理笔记自动归类 tag 标签，然后自动将标签内容整理成阅读笔记
在B 站 UP 主[科研少女王同学](https://space.bilibili.com/686683222)和[[Item/Text] 一个可以关联注释的模板](https://github.com/windingwind/zotero-better-notes/discussions/1101?sort=new#top)
#1101 的这个版本上进行了一些修改。

### Screenshots

<img width="1010" alt="截屏2024-10-18 13 01 35" src="https://github.com/user-attachments/assets/4a3124cc-a71e-4ded-a314-893037a6f094">
<img width="431" alt="截屏2024-10-18 13 01 49" src="https://github.com/user-attachments/assets/4f1c0c42-cb2a-4f66-b2fa-00f5dd57d784">

### Test on Zotero version

最新 7.07 版本

### Test on Better Notes version

2.0.16

### Template Type

QuickNote

### Template Share Code

`${await new Promise(async (r) => {
 const getAnnotationsByTag = async (tagName) => {
  const pdfItem = await topItem.getBestAttachment();
  let result = "";
  for (let annoItem of pdfItem.getAnnotations()) {
   if (annoItem.getTags().find(i => i.tag.includes(tagName))) {
 
    const res = Zotero.EditorInstanceUtilities.serializeAnnotations(
     [
      {
       ...(await Zotero.Annotations.toJSON(annoItem)), ...{ attachmentItemID: annoItem.parentID }
      }
     ]
    );
    result += res.html;
   }
  }
  return result;
 };
 
 // 辅助函数：检测字符串是否包含中文字符
 const containsChinese = (text) => /[\u4e00-\u9fa5]/.test(text);
 
 // 提取影响因子
 let impactFactor = Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").textContent.trim();
 let impactFactorValues = impactFactor.match(/\d+(\.\d+)?/g); // 提取数字部分
 
 // 获取翻译标题，如果没有或者为空，则使用原始标题
 let titleTranslationMatch = topItem.getField("extra").match(/titleTranslation:(.+)/);
 let titleTranslation = (titleTranslationMatch && titleTranslationMatch[1].trim()) ? titleTranslationMatch[1].trim() : topItem.getField("title");
 
 // 检查标题是否为中文
 let isChineseTitle = containsChinese(topItem.getField("title"));
 
 // 根据标题语言决定IFQ的显示方式
 let impactFactorDisplay = isChineseTitle ? impactFactor : (impactFactorValues ? impactFactorValues.join(' / ') : '');
 
 // 获取作者显示内容
 let creators = topItem.getCreators();
 let authorDisplay = "";
 if (isChineseTitle) {
  // 中文标题：显示第一作者全名，格式为“姓 前名”
  if (creators.length > 0) {
   let firstAuthor = creators[0];
   let fullName = `${firstAuthor.lastName} ${firstAuthor.firstName}`;
   authorDisplay = fullName;
  }
 } else {
  // 非中文标题：显示第一作者姓氏并附加“et al.”
  if (creators.length > 0) {
   let firstAuthorLastName = creators[0].lastName;
   authorDisplay = `${firstAuthorLastName} et al.`;
  }
 }
 
 // 生成标题，处理中文和英文情况
 let res = `<h1>🔤${titleTranslation} (<span style="background-color: #B1A5E1">${topItem.getField("year")}</span>, <span style="background-color: #CBE1A5">${authorDisplay}</span>) ${topItem.getField("journalAbbreviation")} (${impactFactorDisplay})</h1>
<p><strong><span style="color: #283593">原名：</span></strong>${topItem.getField('title')}</p>
<p><strong><span style="color: #283593">译名：</span></strong>${titleTranslation}</p>
<p><strong><span style="color: #283593">作者：</span></strong>${authorDisplay}</p>
<p><strong><span style="color: #283593">期刊：</span></strong>${topItem.getField('publicationTitle')}</p>
<p><strong><span style="color: #283593">IFQ：</span></strong><span style="background-color: #CBE1A5">${impactFactorDisplay}</span></p>
 
${(() => {
  const doi = topItem.getField("DOI");
  if (doi) {
   return `<b><span style="color: #283593">DOI：</span></strong> </b><a href="https://doi.org/${topItem.getField('DOI')}">${topItem.getField('DOI')}</a>`;
  } else {
   return `<b><span style="color: #283593">URL：</span></strong> </b><a href="${topItem.getField('url')}">${topItem.getField('url')}</a>`;
  }
 })()}
<p><strong><span style="color: #283593">发表时间：</span></strong>${topItem.getField('date')}</p>
 

<tr>
  <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
    ${(() => {
     const attachments = Zotero.Items.get(topItem.getAttachments());
     if (attachments && attachments.length > 0) {
      return `<b><span style="color: #283593">本地链接:</span> </b><a href="zotero://open-pdf/0_${attachments[0].key}">${attachments[0].getFilename()}</a>`;
     } else {
      return `<b><span style="color: #283593">本地链接:</span> </b>`;
     }
    })()}
  </td>
</tr>
</p>

<tr>
  <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
    ${(() => {
     const abstractTranslation = topItem.getField('abstractTranslation');
     if (abstractTranslation) {
      return `<b><span style="color: #283593">摘要翻译：</span> </b><i>${abstractTranslation}</i>`;
     } else {
      return `<b><span style="color: #283593">摘要：</span> </b><i>${topItem.getField('abstractNote')}</i>`;
     }
    })()}
  </td>
</tr>
 
<h2>💡Keynotes</h2>
<h3><span style="background-color: #d5dbcc">${topItem.getField('journalAbbreviation')} (${impactFactorDisplay}): ${titleTranslation}；${topItem.getField('date')}</span></h3>
<p>${await getAnnotationsByTag('Keynotes')}</p>

<h4><span style="background-color: #a4c7ff">💬 background：</span></h4>
<p> ${await getAnnotationsByTag('background')} </p >
 
<h4><span style="background-color: #ff43ae">🗝️ method：</span></h4>
<p> ${await getAnnotationsByTag('method')} </p >
 
 
<h4><span style="background-color: #1ac469">🥳 conclusion：</span></h4>
<p>${await getAnnotationsByTag('conclusion')}</p >
<p></p >
 
<h4><span style="background-color: #febaed">🔬innovation：</span></h4>
<p>${await getAnnotationsByTag('innovation')}</p >
 
<h4><span style="background-color: #ffd058">📌 shortage：</span></h4>
<p>${await getAnnotationsByTag('shortage')}</p >

<h4><span style="background-color: #9f7cef">🧐 question：</span></h4>
<p>${await getAnnotationsByTag('question')}</p >

<h4><span style="background-color: #ff6643">📝 somethingNew：</span></h4>
<p>${await getAnnotationsByTag('somethingNew')}</p >

<h4><span style="background-color: #afffe3">🗂️ case study：</span></h4>
<p>${await getAnnotationsByTag('case study')}</p >

<h4><span style="background-color: #ddff00">📊 evidence analyse：</span></h4>
<p>${await getAnnotationsByTag('evidence analyse')}</p >
 
`;
 
 r(res);
})}
`

**Comment by windingwind** [comment]
建议使用模板编辑器下方选项-复制模板分享码，方便其他用户直接复制导入

**Reply by swimpinkNo1** [reply]
不好意思，第一次发帖子，不明白在哪里能更方便用户复制导入。😓

**Reply by windingwind** [reply]
https://github.com/windingwind/zotero-better-notes/blob/master/docs/about-note-template.md#share-note-template

模板编辑器下方，选项按钮-复制模板分享码菜单

**Reply by windingwind** [reply]
<img width="956" alt="image" src="https://github.com/user-attachments/assets/a3567ed5-4e7e-4dcd-804b-6f2c876d7f7e">

**Comment by swimpinkNo1** [comment]
求各位大神能够在优化一下笔记模板，在再次阅读文献和编辑注释的时候，笔记可以同步自动更新和修改？

**Reply by windingwind** [reply]
不能够自动，可以使用编辑器下方的 启用模板内容更新 快捷按钮，然后在笔记菜单中手动刷新内容

**Comment by konglong1128** [comment]
up主你好，有这个模板的代码可以分享一下吗

**Reply by swimpinkNo1** [reply]
# This template is specifically for importing/sharing, using better 
# notes 'import from clipboard': copy the content and
# goto Zotero menu bar, click Tools->New Template from Clipboard.  
# Do not copy-paste this to better notes template editor directly.
name: "[item]keynotes"
zoteroVersion: "7.0.7"
pluginVersion: "2.0.18"
savedAt: "2024-10-31T13:28:34.193Z"
content: |-
  ${await new Promise(async (r) => {
   const getAnnotationsByTag = async (tagName) => {
    const pdfItem = await topItem.getBestAttachment();
    let result = "";
    for (let annoItem of pdfItem.getAnnotations()) {
     if (annoItem.getTags().find(i => i.tag.includes(tagName))) {
   
      const res = Zotero.EditorInstanceUtilities.serializeAnnotations(
       [
        {
         ...(await Zotero.Annotations.toJSON(annoItem)), ...{ attachmentItemID: annoItem.parentID }
        }
       ]
      );
      result += res.html;
     }
    }
    return result;
   };
   
   // 辅助函数：检测字符串是否包含中文字符
   const containsChinese = (text) => /[\u4e00-\u9fa5]/.test(text);
   
   // 提取影响因子
   let impactFactor = Zotero.ZoteroStyle.api.renderCell(topItem, "publicationTags").textContent.trim();
   let impactFactorValues = impactFactor.match(/\d+(\.\d+)?/g); // 提取数字部分
   
   // 获取翻译标题，如果没有或者为空，则使用原始标题
   let titleTranslationMatch = topItem.getField("extra").match(/titleTranslation:(.+)/);
   let titleTranslation = (titleTranslationMatch && titleTranslationMatch[1].trim()) ? titleTranslationMatch[1].trim() : topItem.getField("title");
   
   // 检查标题是否为中文
   let isChineseTitle = containsChinese(topItem.getField("title"));
   
   // 根据标题语言决定IFQ的显示方式
   let impactFactorDisplay = isChineseTitle ? impactFactor : (impactFactorValues ? impactFactorValues.join(' / ') : '');
   
   // 获取作者显示内容
   let creators = topItem.getCreators();
   let authorDisplay = "";
   if (isChineseTitle) {
    // 中文标题：显示第一作者全名，格式为“姓 前名”
    if (creators.length > 0) {
     let firstAuthor = creators[0];
     let fullName = `${firstAuthor.lastName} ${firstAuthor.firstName}`;
     authorDisplay = fullName;
    }
   } else {
    // 非中文标题：显示第一作者姓氏并附加“et al.”
    if (creators.length > 0) {
     let firstAuthorLastName = creators[0].lastName;
     authorDisplay = `${firstAuthorLastName} et al.`;
    }
   }
   
   // 生成标题，处理中文和英文情况
   let res = `<h1>🔤${titleTranslation} (<span style="background-color: #B1A5E1">${topItem.getField("year")}</span>, <span style="background-color: #CBE1A5">${authorDisplay}</span>) ${topItem.getField("journalAbbreviation")} (${impactFactorDisplay})</h1>
  <p><strong><span style="color: #283593">原名：</span></strong>${topItem.getField('title')}</p>
  <p><strong><span style="color: #283593">译名：</span></strong>${titleTranslation}</p>
  <p><strong><span style="color: #283593">作者：</span></strong>${authorDisplay}</p>
  <p><strong><span style="color: #283593">期刊：</span></strong>${topItem.getField('publicationTitle')}</p>
  <p><strong><span style="color: #283593">IFQ：</span></strong><span style="background-color: #CBE1A5">${impactFactorDisplay}</span></p>
   
  ${(() => {
    const doi = topItem.getField("DOI");
    if (doi) {
     return `<b><span style="color: #283593">DOI：</span></strong> </b><a href="https://doi.org/${topItem.getField('DOI')}">${topItem.getField('DOI')}</a>`;
    } else {
     return `<b><span style="color: #283593">URL：</span></strong> </b><a href="${topItem.getField('url')}">${topItem.getField('url')}</a>`;
    }
   })()}
  <p><strong><span style="color: #283593">发表时间：</span></strong>${topItem.getField('date')}</p>
   
  
  <tr>
    <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
      ${(() => {
       const attachments = Zotero.Items.get(topItem.getAttachments());
       if (attachments && attachments.length > 0) {
        return `<b><span style="color: #283593">本地链接:</span> </b><a href="zotero://open-pdf/0_${attachments[0].key}">${attachments[0].getFilename()}</a>`;
       } else {
        return `<b><span style="color: #283593">本地链接:</span> </b>`;
       }
      })()}
    </td>
  </tr>
  </p>
  
  <tr>
    <td style="color:#193c47; background-color:#dbeedd; padding:8px;">
      ${(() => {
       const abstractTranslation = topItem.getField('abstractTranslation');
       if (abstractTranslation) {
        return `<b><span style="color: #283593">摘要翻译：</span> </b><i>${abstractTranslation}</i>`;
       } else {
        return `<b><span style="color: #283593">摘要：</span> </b><i>${topItem.getField('abstractNote')}</i>`;
       }
      })()}
    </td>
  </tr>
   
  <h2>💡Keynotes</h2>
  <h3><span style="background-color: #d5dbcc">${topItem.getField('journalAbbreviation')} (${impactFactorDisplay}): ${titleTranslation}；${topItem.getField('date')}</span></h3>
  <p>${await getAnnotationsByTag('Keynotes')}</p>
  
  <h4><span style="background-color: #a4c7ff">💬 background：</span></h4>
  <p> ${await getAnnotationsByTag('background')} </p >
   
  <h4><span style="background-color: #ff43ae">🗝️ method：</span></h4>
  <p> ${await getAnnotationsByTag('method')} </p >
   
   
  <h4><span style="background-color: #1ac469">🥳 conclusion：</span></h4>
  <p>${await getAnnotationsByTag('conclusion')}</p >
  <p></p >
   
  <h4><span style="background-color: #febaed">🔬innovation：</span></h4>
  <p>${await getAnnotationsByTag('innovation')}</p >
   
  <h4><span style="background-color: #ffd058">📌 shortage：</span></h4>
  <p>${await getAnnotationsByTag('shortage')}</p >
  
  <h4><span style="background-color: #9f7cef">🧐 question：</span></h4>
  <p>${await getAnnotationsByTag('question')}</p >
  
  <h4><span style="background-color: #ff6643">📝 somethingNew：</span></h4>
  <p>${await getAnnotationsByTag('somethingNew')}</p >
  
  <h4><span style="background-color: #afffe3">🗂️ case study：</span></h4>
  <p>${await getAnnotationsByTag('case study')}</p >
  
  <h4><span style="background-color: #ddff00">📊 evidence analyse：</span></h4>
  <p>${await getAnnotationsByTag('evidence analyse')}</p >
   
  `;
   
   r(res);
  })}
